From d06cdc04a15ba913d64d04d118486dc02ab8f190 Mon Sep 17 00:00:00 2001
From: Peter Jones <pjones@redhat.com>
Date: Thu, 20 Nov 2014 13:45:25 -0500
Subject: [PATCH 5/5] Add current OsIndications values.

Signed-off-by: Peter Jones <pjones@redhat.com>
---
 inc/efidef.h | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/inc/efidef.h b/inc/efidef.h
index 666b193..8b70051 100644
--- a/inc/efidef.h
+++ b/inc/efidef.h
@@ -208,4 +208,13 @@ typedef UINT8   ISO_639_2;
 #define EFI_SIZE_TO_PAGES(a)  \
     ( ((a) >> EFI_PAGE_SHIFT) + ((a) & EFI_PAGE_MASK ? 1 : 0) )
 
+#define EFI_OS_INDICATIONS_BOOT_TO_FW_UI        0x0000000000000001
+#define EFI_OS_INDICATIONS_TIMESTAMP_REVOCATION 0x0000000000000002
+#define EFI_OS_INDICATIONS_FILE_CAPSULE_DELIVERY_SUPPORTED \
+                                                0x0000000000000004
+#define EFI_OS_INDICATIONS_FMP_CAPSULE_SUPPORTED \
+                                                0x0000000000000008
+#define EFI_OS_INDICATIONS_CAPSULE_RESULT_VAR_SUPPORTED \
+                                                0x0000000000000010
+
 #endif
-- 
2.1.0

