From 6671a712db5c4dfecd4e7026057a99d9b1c4fe5e Mon Sep 17 00:00:00 2001
From: Peter Jones <pjones@redhat.com>
Date: Tue, 7 Mar 2017 10:11:19 -0500
Subject: [PATCH] Nerf -Werror=pragma away

---
 Make.defaults | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/Make.defaults b/Make.defaults
index a7778a7..8882701 100755
--- a/Make.defaults
+++ b/Make.defaults
@@ -87,6 +87,8 @@ OBJDIR       := $(TOPDIR)/$(ARCH)
 # Arch-specific compilation flags
 CPPFLAGS += -DCONFIG_$(ARCH)
 
+CFLAGS += -Wno-error=pragmas
+
 ifeq ($(ARCH),ia64)
   CFLAGS += -mfixed-range=f32-f127
 endif
-- 
2.9.3

