diff -up gnu-efi-3.0d/lib/x86_64/callwrap.c.call10 gnu-efi-3.0d/lib/x86_64/callwrap.c
--- gnu-efi-3.0d/lib/x86_64/callwrap.c.call10	2008-03-27 13:37:37.000000000 -0400
+++ gnu-efi-3.0d/lib/x86_64/callwrap.c	2008-03-27 13:38:08.000000000 -0400
@@ -307,6 +307,7 @@ static UINT64 LIN2WIN10(
 		"movq %[rarg7], " lin2win_win_arg(7,10) "\n\t"
 		"movq %[rarg8], " lin2win_win_arg(8,10) "\n\t"
 		"movq %[rarg9], " lin2win_win_arg(9,10) "\n\t"
+		"movq %[rarg10], " lin2win_win_arg(10,10) "\n\t"
 		alloc_win_stack_frame(10)
 		"call *%[fptr]\n\t"
 		free_win_stack_frame(10)
diff -up gnu-efi-3.0d/inc/x86_64/efibind.h.call10 gnu-efi-3.0d/inc/x86_64/efibind.h
