diff -urpN gnu-efi-3.0d/gnuefi/crt0-efi-ia32.S.attribs gnu-efi-3.0d/gnuefi/crt0-efi-ia32.S
--- gnu-efi-3.0d/gnuefi/crt0-efi-ia32.S.attribs	2008-04-29 13:56:27.000000000 -0400
+++ gnu-efi-3.0d/gnuefi/crt0-efi-ia32.S	2008-04-29 13:56:29.000000000 -0400
@@ -56,7 +56,7 @@ _start:
 dummy:	.long	0
 
 #define IMAGE_REL_ABSOLUTE	0
- 	.section .reloc, "a"
+ 	.section .reloc
  	.long	dummy					// Page RVA
  	.long	10					// Block Size (2*4+2)
  	.word	(IMAGE_REL_ABSOLUTE<<12) +  0		// reloc for dummy
