diff -up gnu-efi-3.0d/apps/Makefile.rpm gnu-efi-3.0d/apps/Makefile
diff -up gnu-efi-3.0d/Make.defaults.rpm gnu-efi-3.0d/Make.defaults
--- gnu-efi-3.0d/Make.defaults.rpm	2007-05-11 13:03:05.000000000 -0400
+++ gnu-efi-3.0d/Make.defaults	2007-12-19 15:08:35.000000000 -0500
@@ -26,6 +26,7 @@
 # lib and include under the root
 #
 INSTALLROOT=/usr/local
+LIBDIR=lib
 
 TOPDIR    := $(shell if [ "$$PWD" != "" ]; then echo $$PWD; else pwd; fi)
 
@@ -78,9 +79,9 @@ else
   LD           = $(prefix)ld
   AR           = $(prefix)ar
   RANLIB       = $(prefix)ranlib
-# OBJCOPY of binutils with x86_64 support
-  OBJCOPY      = /usr/local/bin/objcopy
+  OBJCOPY      = $(prefix)objcopy
   CFLAGS       += -DEFI_FUNCTION_WRAPPER
+  LIBDIR       = lib64
  endif
 
  endif
diff -up gnu-efi-3.0d/gnuefi/Makefile.rpm gnu-efi-3.0d/gnuefi/Makefile
--- gnu-efi-3.0d/gnuefi/Makefile.rpm	2007-12-19 15:09:29.000000000 -0500
+++ gnu-efi-3.0d/gnuefi/Makefile	2007-12-19 15:09:41.000000000 -0500
@@ -37,8 +37,8 @@ clean:
 	rm -f $(TARGETS) *~ *.o $(OBJS)
 
 install:
-	mkdir -p $(INSTALLROOT)/lib
-	$(INSTALL) -m 644 $(TARGETS) $(INSTALLROOT)/lib
-	$(INSTALL) -m 644 elf_$(ARCH)_efi.lds $(INSTALLROOT)/lib
+	mkdir -p $(INSTALLROOT)/$(LIBDIR)
+	$(INSTALL) -m 644 $(TARGETS) $(INSTALLROOT)/$(LIBDIR)
+	$(INSTALL) -m 644 elf_$(ARCH)_efi.lds $(INSTALLROOT)/$(LIBDIR)
 
 include ../Make.rules
diff -up gnu-efi-3.0d/lib/Makefile.rpm gnu-efi-3.0d/lib/Makefile
--- gnu-efi-3.0d/lib/Makefile.rpm	2007-12-19 15:07:48.000000000 -0500
+++ gnu-efi-3.0d/lib/Makefile	2007-12-19 15:08:43.000000000 -0500
@@ -45,7 +45,7 @@ clean:
 	rm -f libefi.a *~ $(OBJS) */*.o
 
 install: libefi.a
-	mkdir -p $(INSTALLROOT)/lib
-	$(INSTALL) -m 644 libefi.a $(INSTALLROOT)/lib/
+	mkdir -p $(INSTALLROOT)/$(LIBDIR)
+	$(INSTALL) -m 644 libefi.a $(INSTALLROOT)/$(LIBDIR)/
 
 include ../Make.rules
